### PR TITLE
fix(qwen3-vl): honor enable_precise_embedding_interpolation on the default ViT path

### PR DESCRIPTION
**Pull request by Hert4**
## Motivation

`enable_precise_embedding_interpolation` (`align_corners`) has **no effect on the
default Qwen3-VL ViT path**. Both branches of the eager forward interpolate the ViT
absolute position-embedding grid with `linspace` coordinates unconditionally, while
the CUDA-graph path honors the flag — and it even carries the exact guard the eager
path is missing.

Current `main`, `python/sglang/srt/models/qwen3_vl.py`:

```python
# eager path (default: SGLANG_VIT_ENABLE_CUDA_GRAPH is off)
903:  pos_embeds = self.fast_pos_embed_interpolate_vectorized(grid_thw_list)  # linspace
905:  pos_embeds = self.fast_pos_embed_interpolate_from_list(grid_thw_list)   # linspace

# CUDA-graph path
1085: if self.align_corners and self._use_vectorized_pos_embed(len(grid_thw_list)):
          # The vectorized implementation uses linspace coordinates. In graph mode
          # the legacy fallback honors enable_precise_embedding_interpolation, so
          # only use the vectorized path when the active graph interpolation mode
          # is also linspace; otherwise image count would change the output.
1090:     pos_embeds = self.fast_pos_embed_interpolate_vectorized(grid_thw_list)
1092: else:
          pos_embeds = self.fast_pos_embed_interpolate(grid_thw)               # honors flag
```

`fast_pos_embed_interpolate_from_list()` hardcodes the coordinates:

```python
547: def fast_pos_embed_interpolate_from_list(self, grid_thw):
554:     h_idxs = torch.linspace(0, num_grid_per_side - 1, h, ...)   # == align_corners=True
557:     w_idxs = torch.linspace(0, num_grid_per_side - 1, w, ...)
```

whereas the flag-honoring helpers pick half-pixel centers by default:

```python
461: def _get_interpolation_indices(self, dim_size):        # numpy, used by the graph path
467:     if self.align_corners:
468:         indices = np.linspace(0, self.num_grid_per_side - 1, dim_size, ...)
471:     else:
472:         indices = (np.arange(dim_size) + 0.5) * (self.num_grid_per_side / dim_size) - 0.5
475:         indices = np.clip(indices, 0, self.num_grid_per_side - 1)

531: def _torch_interp_indices(self, dim_size, device):     # torch mirror, currently unused
541:     # align_corners=False  (match _get_interpolation_indices)
```

`_torch_interp_indices()` already exists and documents itself as matching
`_get_interpolation_indices()`, but nothing calls it — this PR wires it up.

Two more signals that all paths were meant to agree:

- `_use_vectorized_pos_embed()` docstring: *"Both give the same result."*
- `fast_pos_embed_interpolate_vectorized()` docstring: *"Same result as the loop version."*

Both statements hold between the two `linspace` implementations, but not against
`fast_pos_embed_interpolate()` at the default flag value.

## Secondary: the flag-honoring path is unreachable on a default Hopper setup

`fast_pos_embed_interpolate()` is only reached through the ViT CUDA-graph path, and
that path aborts with the default vision attention backend. On H200, sglang
`0.0.0.dev1+ga14971730.d20260731`, `Qwen3.5-35B-A3B-FP8`, TP1, `attention_backend=fa3`,
`mm_attention_backend=None`, setting `SGLANG_VIT_ENABLE_CUDA_GRAPH=1` takes the
scheduler down on the first image request:

```
File "python/sglang/srt/models/qwen3_vl.py", line 1032, in forward_with_cuda_graph
    return self.graph_runners.run(
File "python/sglang/srt/multimodal/vit_cuda_graph_runner.py", line 200, in _create_graph
    raise RuntimeError("Not supported ViT attention backend")
[...] SIGQUIT received. signum=None [...] It usually means one child failed.
[...] "POST /v1/chat/completions HTTP/1.1" 500 Internal Server Error
```

So in practice every request takes the eager path, and
`enable_precise_embedding_interpolation` cannot be made to take effect at all.
Arguably the runner should fall back to eager (or fail at startup) instead of
`SIGQUIT`-ing on the first request, but that is a separate change and not part of
this PR.

## Modifications

- `fast_pos_embed_interpolate_from_list()`: use `_torch_interp_indices()` instead of
  hardcoded `torch.linspace(...)`, so the default path honors the flag.
- New unit test pinning both interpolation paths to the same coordinates for both
  flag values, plus a test asserting the flag is observable on the default path.
  Registered with `register_cpu_ci` (no GPU, no weights needed).

## Alternatives

The asymmetry can be resolved in either direction and I have no strong preference —
happy to switch:

1. **This PR**: make the eager path honor the flag (default becomes half-pixel
   centers, matching `_get_interpolation_indices`' documented default).
2. Make the graph path use `linspace` unconditionally too, and document that
   `enable_precise_embedding_interpolation` no longer applies to Qwen3-VL.

Option 1 changes default numerics for Qwen3-VL/Qwen3.5/Moss-VL, which is why I am
flagging it explicitly rather than presenting it as a pure no-op cleanup. Note that
`fast_pos_embed_interpolate_vectorized()` is also `linspace`-only; it is gated behind
`SGLANG_VIT_ENABLE_VECTORIZED_POS_EMBED` and `num_images >= N`, so it is untouched
here, but it has the same inconsistency and should follow whichever direction is
chosen.

## Testing

```
pytest test/registered/unit/models/test_qwen3_vl_pos_embed_interp.py -v
```

The tests construct only the attributes the two paths touch, so they need no model
weights, no GPU, and no vision config.

Before this change, `test_from_list_matches_reference_path[grid...-False]` and
`test_flag_actually_changes_from_list_output` fail; after it, all 13 pass.

## Not claimed

I am **not** claiming a quality regression from this. I measured a ~4 point
field-accuracy difference between sglang `0.5.9` and `20260731` nightly on an
internal document-extraction task, but a controlled experiment ruled this
interpolation path out as its cause, and that dataset is too small (17 documents)
to support a magnitude claim anyway. This PR is a consistency fix only.

## Checklist

- [x] Format your code with `pre-commit run --files ...`
- [x] Add unit tests

---
### CI States

Latest PR Test (Base): :x: [Run #31206271570](https://github.com/sgl-project/sglang/actions/runs/31206271570)
Latest PR Test (Extra): :x: [Run #31206270548](https://github.com/sgl-project/sglang/actions/runs/31206270548)

### CODE REVIEW

**Comment by Hert4** [comment]
cc @yuan-luo @JustinTong0323, this wires `_torch_interp_indices()` (introduced in #19003) into `fast_pos_embed_interpolate_from_list()` so the default eager path honors `enable_precise_embedding_interpolation` the same way the ViT CUDA-graph path already does.

If you'd rather resolve the asymmetry the other way (make everything `linspace` and retire the flag for Qwen3-VL — option 2 in the PR body), happy to switch, it's a one-line change.

Could you tag `run-ci` when you get a chance? The new test is CPU-only (`base-a-test-cpu`), no GPU needed.

**Comment by Hert4** [comment]
### When did the divergence start: `v0.5.10rc0` / `v0.5.10`, from #19003

Since "how long has this been broken" is the obvious first review question, I bisected it.

Method: classified `python/sglang/srt/models/qwen3_vl.py` at **every** release tag from `v0.4.0`
to `v0.5.17` (blob at tag, not merge dates), then pinned the introducing commits with `git log -S`
on a full clone. Blob-at-tag, `git log -S` and `git merge-base --is-ancestor` all agree.

| tag range | eager `forward()` calls | is the flag effective there? |
|---|---|---|
| < `v0.5.3rc2` | file does not exist | — |
| `v0.5.3rc2` … `v0.5.6.post2` | `fast_pos_embed_interpolate` — `np.linspace`, unconditional | n/a, flag does not exist yet |
| **`v0.5.7` … `v0.5.9`** | `fast_pos_embed_interpolate` — honors `align_corners` | ✅ **yes** |
| **`v0.5.10rc0` … `v0.5.17` (current main)** | `fast_pos_embed_interpolate_from_list` — hardcoded `linspace` | ❌ **no** |
| `v0.5.15`+ | additionally `_vectorized` behind `SGLANG_VIT_ENABLE_VECTORIZED_POS_EMBED` (also `linspace`-only) | ❌ no |

So `v0.5.9` and earlier are **not** affected — there both `forward()` and the graph path went
through the same flag-honoring helper.

**Introduced by #19003** — *"[VLM] Introduce FlashInfer CUDNN Prefill as ViT Backend"*,
commit `31c7dc9d99`, merged 2026-02-24, first released in `v0.5.10rc0` (2026-03-27) and
`v0.5.10` (2026-04-05). That commit added `fast_pos_embed_interpolate_from_list()` and pointed the
eager `forward()` at it while `_prepare_graph_inputs()` kept the flag-honoring helper. It also
added `_torch_interp_indices()` and never called it — the dead helper and the divergence arrive in
the same commit, which is why wiring it up is the minimal fix.

Call sites, verified per tag:

```
v0.5.9      L543 fast_pos_embed_interpolate            <- def forward()
            L595 fast_pos_embed_interpolate            <- def forward_with_cuda_graph()
v0.5.10rc0  L751 fast_pos_embed_interpolate_from_list  <- def forward()             <-- divergence
            L932 fast_pos_embed_interpolate            <- def _prepare_graph_inputs()
v0.5.17     L903/L905 _vectorized / _from_list         <- def forward()
            L1090/L1092                                <- def _prepare_graph_inputs()
```

Full provenance:

| what | commit | PR | first tag |
|---|---|---|---|
| flag + `align_corners` (via `F.interpolate`) | `1048803c1f` | #10959 | `v0.5.7` |
| `_get_interpolation_indices()` (numpy index form) | `6a4b81e2d9` | #16781 | `v0.5.9` |
| `fast_pos_embed_interpolate_from_list()` + dead `_torch_interp_indices()` | `31c7dc9d99` | #19003 | `v0.5.10rc0` |
| `fast_pos_embed_interpolate_vectorized()` (also `linspace`-only) | `0df796473b` | #28940 | `v0.5.15` |

Flag reachability was checked per tag rather than assumed: `v0.5.7`/`v0.5.8`/`v0.5.8.post1` pass
`align_corners=self.align_corners` straight into `F.interpolate`; `v0.5.9` calls
`self._get_interpolation_indices()` from inside `fast_pos_embed_interpolate()`. The server-arg
default is `False` at every tag checked (`v0.5.7`, `v0.5.9`, `v0.5.10`, `v0.5.12`, `v0.5.15`,
`v0.5.17`), and the wiring is direct: `self.align_corners = <flag>`.

---

### This changes my recommendation: I now think Alternative 2 is the right direction

Upstream `transformers` hardcodes **`interpolation_align_corners = True`** for Qwen3-VL
(`modeling_qwen3_vl.py:627` on `main`, and the original `v4.57.0` release used plain
`torch.linspace`). So `linspace` — i.e. `align_corners=True` — is the reference behavior, and the
default numerics here have flipped twice:

| sglang | default eager coordinates | matches the HF reference? |
|---|---|---|
| ≤ `v0.5.6.post2` | `linspace` | ✅ |
| **`v0.5.7` – `v0.5.9`** | half-pixel centers (flag defaults to `False`) | ❌ deviates |
| ≥ `v0.5.10` | `linspace` (flag ignored) | ✅, accidentally |

Alternative 1 as written in the PR description makes the default half-pixel again, which
re-introduces the `v0.5.7`–`v0.5.9` deviation from `transformers` on the default path. That is not
what I want. Two options that keep the default reference-matching:

- **Alternative 2** — `linspace` everywhere, document that
  `enable_precise_embedding_interpolation` does not apply to Qwen3-VL. Pure consistency fix, zero
  numerics change.
- Keep this PR's wiring (so the flag is genuinely honored on all paths) **and** flip the flag
  default to `True`, so the out-of-the-box path stays on the HF-reference coordinates and the flag
  becomes an opt-*out* of it.

I have a mild preference for the second, since it makes `_torch_interp_indices()` live and the
flag meaningful, but I am happy to push whichever maintainers prefer — say the word and I will
update the PR.

For scale, with `num_grid_per_side = 48`, the two coordinate schemes differ by up to 0.31 grid
cells on large grids and are exactly equal at `h = 48`. On a smooth learned-like 48×48 table the
resulting position embedding differs by roughly 0.6–2% relative L2 at document-page grid sizes
(`46×62` → `128×96`), rising to ~21% at `8×8`, concentrated at the image borders.

This also locates the `0.5.9` vs `20260731`-nightly pair I mentioned in the description: that pair
straddles the `v0.5.9`/`v0.5.10` flip, with the nightly on the HF-matching side. As stated there, a
controlled experiment already ruled this path out as the cause of the accuracy difference I saw,
and the magnitudes above are consistent with that — I am not claiming a quality regression here.

**Comment by Hert4** [comment]
@yuan-luo @mickqian could one of you add the `run-ci` label?
